In [1]:
2 + 1

3

In [2]:
2 - 1

1

In [3]:
2 * 2

4

In [4]:
2 / 3

0.6666666666666666

Modulo

In [6]:
7 % 4

3

In [9]:
(-7) % 3

2

Power

In [10]:
2 ** 3

8

`https://docs.python.org/2/tutorial/floatingpoint.html`
No matter how many base 2 digits you’re willing to use, the decimal value 0.1 cannot be represented exactly as a base 2 fraction. In base 2, 1/10 is the infinitely repeating fraction. On a typical machine running Python, there are 53 bits of precision available for a Python float. 

It’s easy to forget that the stored value is an approximation to the original decimal fraction, because of the way that floats are displayed at the interpreter prompt.

In [12]:
0.1+0.2

0.30000000000000004

Other surprises follow from this one. 
For example, if you try to round the value 2.675 to two decimal places, you get this:

In [14]:
round(2.675, 2)

2.67

Since the decimal fraction 2.675 is exactly halfway between 2.67 and 2.68, you might expect the result here to be (a binary approximation to) 2.68. It’s not, because when the decimal string 2.675 is converted to a binary floating-point number, it’s again replaced with a binary approximation, whose exact value is 2.67499999999999982236431605997495353221893310546875. 

If you’re in a situation where you care which way your decimal halfway-cases are rounded, you should consider using the decimal module, which provides a nice way to “see” the exact value that’s stored in any particular Python float:

In [15]:
from decimal import Decimal
Decimal(2.675)

Decimal('2.67499999999999982236431605997495353221893310546875')

Still, don’t be unduly wary of floating-point! The errors in Python float operations are inherited from the floating-point hardware, and on most machines are on the order of no more than 1 part in `2**53` per operation. That’s more than adequate for most tasks, but you do need to keep in mind that it’s not decimal arithmetic, and that every float operation can suffer a new rounding error.

While pathological cases do exist, for most casual use of floating-point arithmetic you’ll see the result you expect in the end if you simply round the display of your final results to the number of decimal digits you expect.